ghcide/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -141,7 +141,10 @@ runLanguageServer options inH outH getHieDbLoc defaultConfig onConfigurationChan
             logInfo (ideLogger ide) $ T.pack $ "Registering ide configuration: " <> show initConfig
             registerIdeConfiguration (shakeExtras ide) initConfig
 
-            _ <- flip forkFinally (const exitClientMsg) $ runWithDb dbLoc $ \hiedb hieChan -> do
+            let handleServerException e = do
+                    logError (ideLogger ide) (T.pack $ "Unexpected exception in server thread: " <> show e)
+                    exitClientMsg
+            _ <- flip forkFinally handleServerException $ runWithDb dbLoc $ \hiedb hieChan -> do
```

Comment:
`handleServerException` always runs after the action. We should check `e` is `Left SomeException` or `Right ()`.